Describe how ytt works (with pictures)

pivotaljohn opened the pull request:
- link from "About ytt" to continue the initial journey through ytt
docs.

jtigger committed:
Describe how ytt works (with pictures)

- link from "About ytt" to continue the initial journey through ytt
docs.

jtigger committed:
Some edits w/ Helen

jtigger committed:
More notes from Helen

jtigger committed:
Lots of rewording/working based on feedback

- added pointers from "deeper" spots in the docs back to this
  "fundamentals" primer to make it more discoverable.
- include some terse examples to help make concepts more accessible.

jtigger committed:
At the start, make clear what the reader will learn

Responding this feedback:
https://github.com/vmware-tanzu/carvel.dev/pull/95#discussion_r621481185

pivotaljohn commented on content/ytt/docs/latest/how-it-works.md:
```diff
@@ -0,0 +1,96 @@
+
+## Overview
+
+Having a picture of how your inputs are processed in `ytt` can be helpful.
```
> When I see an overview section in docs, I hope to get out of it "what will I learn if I read this", or "am I the intended audience of this"?

Yeah, I really wrestled with this friggin' sentence.  Somehow, I just kept getting stuck at, "It was a dark and stormy night..." 😉 

Thank you for the needs-based framing, it unsticks it...

Just pushed a change incorporating this. 👇🏼

cari-lynn reviewed (approved):
Thanks for taking the time to improve the docs 
+1 for the sql analogy

cppforlife commented on content/ytt/docs/latest/how-it-works.md:
````diff
@@ -0,0 +1,150 @@
+When you invoke `ytt` ...
+
+```console
+$ ytt -f (input files)
````
should this command look a bit more real. it looks a bit confusing since () mean something else in bash.

cppforlife commented on content/ytt/docs/latest/how-it-works.md:
````diff
@@ -0,0 +1,150 @@
+  #@ for/end name in ["Alice", "Bob", "world"]:
+  - #@ "Hello, " + name
+  ```
+- when a template starts with the [`@data/values`](ytt-data-values.md#declaring-and-using-data-values) annotation, it's called a "Data Values":
````
shouldnt this say YAML document instead of template?

pivotaljohn replied:
Yeah.... so there's this needle I'm trying to thread through this string of terms in hopes of making meaningful distinctions between them... and because the word "template" is overloaded, it complicates the story:

"template" can have three (3) meanings:
1. a file containing `ytt` directives
2. a document containing `ytt` directives
3. (with a capital 'T', "Template") a document that's contributing to output (i.e. _not_ a Data Values nor an Overlay)

And the naming of parts goes something like this:
- a file contains a document set
- a file containing annotations is a template (definition 1; not used in this story)
- a document set contains zero or more documents
- a document, if containing annotations is a template (definition 2)
- templates can be categorized into three groups:
  - Data Values
  - Overlays
  - Templates (definition 3)

I've been trying to avoid getting wordy. For example, could use _these_ terms:

1. a file containing `ytt` directives is a "template"
2. a document containing `ytt` directives is a "templated document"
3. a "templated document" contributing to output is a "templated output document"

Folding that back in to the sentence at hand:

```suggestion
- when a templated document starts with the [`@data/values`](ytt-data-values.md#declaring-and-using-data-values) annotation, it's called a "Data Values":
```

It's a little wordier, but maybe the increased density isn't too much? ... and it has the added benefit of foreshadowsingterms that have meaning (e.g. marking a file for output makes it one of those "output documents")...

cppforlife commented on content/ytt/docs/latest/how-it-works.md:
```diff
@@ -0,0 +1,150 @@
+   
+The result is (shown as "Output DocSet" in the diagram, [above](#the-pipeline)) — the finalized set of YAML documents, in memory. Which leaves one last step...
+
+### Step 4: Print
```
instead of calling this print, may be it should be "serializing". one thing we should probably mention is that because we are dealing with structures, it's always going to come out as legitimate yaml.

pivotaljohn replied:
Yeah, I was trying out the verb from the codebase. If the intro can support the would be contributor — without damaging _this_ explanation — it's all the more valuable.

But re-reading it with your observation, I see that "serializing" is going to be more familiar in this context and convey more important things (i.e. the structure will be preserved).

jtigger committed:
Name output stage as "Serialize" not "Print"

... and give a more realistic example of invoking `ytt` to avoid
confusion.

jtigger committed:
Rework terms used to describe inputs

... to be clear about the distinction between a "template" (a kind of file)
and kinds of YAML documents.

Also:
- introduce color to distinguish between file types and also convey how
  they combine (yellow+blue = green).
- provide examples of each kind of input.
- rework the outline
